wechat/handlers.py

```diff
@@ -65,3 +75,170 @@ def check(self):
+
+    def check(self):
+        return self.is_text('抢啥') or self.is_event_click(self.view.event_keys['book_what'])
+
+    def handle(self):
+        lib = WeChatLib(WECHAT_TOKEN, WECHAT_APPID, WECHAT_SECRET)
+        wechat_menu = lib.get_wechat_menu()
+        if len(wechat_menu) < 2:
+            return self.reply_text('暂时没有可抢票的活动')
+        elif len(wechat_menu[1]['sub_button']) == 0:
+            return self.reply_text('暂时没有可抢票的活动')
+        else:
+            acti_list = wechat_menu[1]['sub_button']
+            news_list = []
+            for acti in acti_list:
+                try:
+                    activity = Activity.objects.get(name=acti['name'])
+                    news_list.append({
+                        'Title': activity.name,
+                        'Description': activity.description,
+                        'PicUrl': activity.pic_url,
+                        'Url': self.url_activity_detail(activity.id)
+                    })
+                except:
+                    return self.reply_text('数据库内出现了重名活动，请联系管理员-_-!')
+
+            return self.reply_news(news_list)
+
+
+class BookTicketHandler(WeChatHandler):
+
+    def check(self):
+        return self.is_text_command('抢票') or self.is_event_click(self.view.event_keys['book_header'])
+
+    def handle(self):
+        if self.is_msg_type('event'):
+            """
+            lib = WeChatLib(WECHAT_TOKEN, WECHAT_APPID, WECHAT_SECRET)
+            acti_list = lib.get_wechat_menu()[1]['sub_button']
+            for acti in acti_list:
+                if acti['key'] == self.input['EventKey']:
+                    activity_name = acti['name']
+                    break
+            try:
+                activity = Activity.objects.get(name=activity_name)
+            """
+            activity_id = int(self.input['EventKey'][len(self.view.event_keys['book_header']):])
+            try:
+                activity = Activity.objects.get(id=activity_id)
+            except:
+                return self.reply_text('该活动不存在')
+        elif self.is_msg_type('text'):
+            try:
+                activity_name = self.input['Content'].split()[1]
+            except:
+                return self.reply_text('格式错误0_0 抢票请输入"抢票 活动名"')
+            try:
+                activity = Activity.objects.get(name=activity_name)
+            except:
+                return self.reply_text('活动名错误?_? 该活动不存在或有重名活动')
+
+        if not self.user.student_id:
+            return self.reply_text('你还没绑定学号哟')
+
+        if len(Ticket.objects.filter(activity=activity, student_id=self.user.student_id)) > 0:
+            return self.reply_text('一个人只能抢一张票哦')
+
+        currentTime = int(time.time())
+        if int(time.mktime(activity.book_start.timetuple())) > currentTime:
+            return self.reply_text('抢票尚未开始')
+        elif int(time.mktime(activity.book_end.timetuple())) < currentTime:
+            return self.reply_text('抢票已结束')
+
+        remainTicketsLock.acquire()
```

Comment:
low efficiency here!
server would only serve a client at one time because of the GIL